client/fingerprint/consul.go

```diff
@@ -298,6 +303,60 @@ func (cfs *consulFingerprintState) grpcTLSPort(info agentconsul.Self) (string, b
 	return fmt.Sprintf("%d", int(p)), ok
 }
 
+func (cfs *consulFingerprintState) dnsPort(info agentconsul.Self) (string, bool) {
+	p, ok := info["DebugConfig"]["DNSPort"].(float64)
+	return fmt.Sprintf("%d", int(p)), ok
+}
+
+// dnsAddr fingerprints the Consul DNS address, but only if Nomad can use it
+// usefully to provide an iptables rule to a task
+func (cfs *consulFingerprintState) dnsAddr(logger hclog.Logger) func(info agentconsul.Self) (string, bool) {
+	return func(info agentconsul.Self) (string, bool) {
+
+		// only addresses we can use for an iptables rule from a container to the
+		// host will be fingerprinted
+		isValidForTaskUse := func(addr netip.Addr) (string, bool) {
+			if !addr.IsLoopback() && !addr.IsUnspecified() && addr.IsValid() {
+				return addr.String(), true
+			}
+			return "", false
+		}
+
+		// first try to find an explicitly configured address
+		dnsAddrs, ok := info["DebugConfig"]["DNSAddrs"].([]string)
+		if ok {
+			for _, dnsAddr := range dnsAddrs {
+				dnsAddr = strings.TrimPrefix(dnsAddr, "tcp://")
+				dnsAddr = strings.TrimPrefix(dnsAddr, "udp://")
+
+				parsed, err := netip.ParseAddrPort(dnsAddr)
+				if err != nil {
+					logger.Warn("could not parse Consul addresses.dns config", "value", dnsAddr)
+					return "", false
```

Comment:
if any of the DNSAddrs are unparseable, then this return would prevent us from attempting the bind addr fallback below eh?

Reply:
Yeah the only reason we'd be getting an error here is if the response body format changed in a way we didn't know about, so trying to guess and continue is the wrong approach.

Reply:
ah hm, I see in the [Consul API docs](https://developer.hashicorp.com/consul/api-docs/agent#read-configuration) that

> `DebugConfig` contains the full runtime configuration but its format is subject to change without notice or deprecation.

are you looking to guard against that warning regarding the structure of the API response?  wouldn't the `ok` check on the `["DNSAddrs"].([]string)` type coercion cover that angle?  I suppose it still could change format ("proto://" etc) within the `[]string`s...

I was assuming that perhaps operators could slightly misconfigure Consul in such a way that one address could be invalid, but other(s) being right and proper could let Consul roll with it, and if that's the case, maybe we could too.  if that assumption is wrong though, and a single invalid address casts doubt on the entire effort, then bailing as early as possible as you've done here makes good sense to me.